### PR TITLE
[jit] Add pickler support for Device

### PR DESCRIPTION
**Pull request by driazati**
This PR adds (un)pickling support for `c10::Device`. It also adds `torch.device` as a type annotation for device attributes.

Differential Revision: [D18664421](https://our.internmc.facebook.com/intern/diff/18664421/)

### CODE REVIEW

**Comment by resistor** [comment]
Could we reduce forward-compat problems by splitting this into a reader change that we ship first, and a writer change that we ship second?

**Comment by driazati** [comment]
This doesn't change any existing behavior (someone has to notice this feature and then go and try to serialize a `Device`), so it won't break anything that already exists, so I think it's pretty low risk as is

File: torch/csrc/jit/unpickler.cpp [hunk 1]
```diff
@@ -494,6 +494,15 @@ PickleOpCode Unpickler::readInstruction() {
...
         });
       } else if (module_name == "torch") {
+        if (class_name == "device") {
+          globals_.emplace_back([this] {
+            auto device_string = stack_.back().toTuple()->elements().at(0);
+            stack_.pop_back();
+            stack_.emplace_back(c10::Device(device_string.toStringRef()));
+          });
+          stack_.emplace_back(int64_t(globals_.size() - 1));
+          return opcode;
```
Comment (zdevito) [inline torch/csrc/jit/unpickler.cpp]: The early returns that have accumulated in this branch of the switch statement should be cleaned up. It is not clear to me why some cases break, others early return, and some push the globals reference number into the stack. Ideally, nothing inside of the switch statement should be returning early.  Perhaps put the global logic in its own function for clarity.

**Review by zdevito — APPROVED** [review]
Looks good, comments inline.

**Pull request closed without merging**